[SIL] Add [serialized_for_package] to control package-wide resilience domain in Package-CMO.

elsh opened the pull request:
This PR adds a new attribute [serialized_for_package] to allow 
more fine-control over package-wide resilience domain if 
Package CMO is enabled.

The purpose of the attribute includes:
- Indicates that certain types such as loadable types are
allowed in serialized functions in resiliently built module
if the optimization is enabled, which are otherwise disallowed.
- Used during SIL deserialization to determine whether such
functions are allowed.
- Used to determine if a callee can be inlined into a caller
that's serialized without package-cmo, e.g. with an explicit
annotation like @inlinable, where the callee was serialized
due to package-cmo.

Resolves rdar://127870822&127321129&127308331

elsh committed:
[SIL] Add a new attribute [serialized_for_package] to support
package-wide resilience domain if Package CMO is enabled.

The purpose of the attribute includes:
- Indicates that certain types such as loadable types are
allowed in serialized functions in resiliently built module
if the optimization is enabled, which are otherwise disallowed.
- Used during SIL deserialization to determine whether such
functions are allowed.
- Used to determine if a callee can be inlined into a caller
that's serialized without package-cmo, e.g. with an explicit
annotation like @inlinable, where the callee was serialized
due to package-cmo.

Resolves rdar://127870822

aschwaighofer commented:
We should block inlining a `[serialized] [serialized_for_package]` function into a `[serialized]`-only function until the `[serialized]` attribute is removed. 

Otherwise, you end up with the code originating from a `[serialized_for_package]` function that assumes loadability in a `[serialized]` function (without the `serialized_for_package`) attributed and the verifier will complain.


We achieve the analogous thing for serialized and not-serialized functions here:

https://github.com/apple/swift/blob/8c0726da80dbddbbbf1fc6ba4f0ab39df1ee3e9a/lib/SILOptimizer/Utils/PerformanceInlinerUtils.cpp#L849-L851

https://github.com/apple/swift/blob/8c0726da80dbddbbbf1fc6ba4f0ab39df1ee3e9a/include/swift/SIL/SILFunction.h#L875

elsh commented:
@swift-ci smoke test

slavapestov commented:
> We should block inlining a [serialized] [serialized_for_package] function into a [serialized]-only function until the [serialized] attribute is removed.

I think the bits should be mutually exclusive no? So both should not be set at once. [serialized] is the stonger condition so if its set, it implies the same invariants as [serialized_for_package] but adds more.

elsh committed:
-Add package-cmo check to bypassResilienceInPackage method.
-Fix SILDeclRef getLinkageLimit() for GlobalAccessor to return
Limit::None if bypassResilienceInPackage is enabled.

elsh commented:
@swift-ci smoke test

elsh commented:
@swift-ci smoke test

elsh committed:
Initialize SerializedForPackage bit field

elsh commented:
@swift-ci smoke test

eeckstein commented on include/swift/SIL/SILLinkage.h:
```diff
@@ -162,6 +162,11 @@ enum IsSerialized_t : unsigned char {
   IsSerialized
 };
 
+enum IsSerializedForPackage_t : unsigned char {
```
I agree with Slava. You should combine both enums, e.g.
```
enum IsSerialized_t : unsigned char {
  IsNotSerialized,
  IsSerialized,
  IsSerializedForPackage
};
```

aschwaighofer reviewed (approved):
I think we can serialization_t tri-state enum as a follow up PR that also addresses inlinability

aschwaighofer commented:
we can address ...
